src/Stratis.SmartContracts.Executor.Reflection/Serialization/Prefix.cs

```diff
@@ -0,0 +1,105 @@
+﻿using System;
+using NBitcoin;
+
+namespace Stratis.SmartContracts.Executor.Reflection.Serialization
+{
+    /// <summary>
+    /// The prefix to use when serializing a method parameter. Represents the <see cref="Type"/> of parameter being serialized.
+    /// </summary>
+    public class Prefix
+    {
+        public byte Value { get; }
+
+        public Prefix(byte value)
+        {
+            this.Value = value;
+        }
+
+        public Type Type => this.GetType(this.Value);
+
+        public MethodParameterDataType DataType => (MethodParameterDataType) this.Value;
+
+        public int Length { get; } = 1;
+
+        private Type GetType(byte b)
+        {
+            switch ((MethodParameterDataType)b)
+            {
+                case MethodParameterDataType.Address:
+                    return typeof(Address);
+                case MethodParameterDataType.Bool:
+                    return typeof(bool);
+                case MethodParameterDataType.Byte:
+                    return typeof(byte);
+                case MethodParameterDataType.Char:
+                    return typeof(char);
+                case MethodParameterDataType.String:
+                    return typeof(string);
+                case MethodParameterDataType.Int:
+                    return typeof(int);
+                case MethodParameterDataType.UInt:
+                    return typeof(uint);
+                case MethodParameterDataType.Long:
+                    return typeof(long);
+                case MethodParameterDataType.ULong:
+                    return typeof(ulong);
+                case MethodParameterDataType.UInt160:
+                    return typeof(uint160);
+                case MethodParameterDataType.ByteArray:
+                    return typeof(byte[]);
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(b), b, "Unsupported type");
+            }
+        }
+
+        public static Prefix ForObject(object o)
+        {
+            var type = (byte) GetPrimitiveType(o);
+            return new Prefix(type);
+        }
+
+        private static MethodParameterDataType GetPrimitiveType(object o)
```

Comment:
If the primitives are the same among a few serializers and we end up with many methods like this, I think there is a better approach here. A static readonly list of the primitive types somewhere with references to the required method paramters and serialization etc? Would help us in future with ensuring each type is covered for each serializer rather than copying these long if returns

Not a blocker, just thoughts!